File: src/shapes/Shape.cc
// -*- LSST-C++ -*-

/* 
 * LSST Data Management System
 * Copyright 2008, 2009, 2010 LSST Corporation.
 * 
 * This product includes software developed by the
 * LSST Project (http://www.lsst.org/).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the LSST License Statement and 
 * the GNU General Public License along with this program.  If not, 
 * see <http://www.lsstcorp.org/LegalNotices/>.
 */
 
#include "lsst/pex/exceptions.h"
#include "lsst/pex/logging/Trace.h"
#include "lsst/afw.h"

#include "lsst/meas/algorithms/Shape.h"

namespace pexExceptions = lsst::pex::exceptions;
namespace pexLogging = lsst::pex::logging;
namespace afwImage = lsst::afw::image;

namespace lsst {
namespace meas {
namespace algorithms {

/************************************************************************************************************/
/**
 * @brief the Shape class, to hold object moments and their covariances
 */
/**
 * Return e1 == (m_{xx} - m_{yy})/(m_{xx} + m_{yy})
 */
double Shape::getE1() const {
    double const T  = (_mxx + _myy);      // Trace
    assert(T > 0);
    return (_mxx - _myy)/T;
}

/**
 * Return e2 == 2*m_{xy}/(m_{xx} + m_{yy})
 */
double Shape::getE2() const {
    double const T  = (_mxx + _myy);      // Trace
    assert(T > 0);
    return 2*_mxy/T;
}

/**
 * Return the object's RMS size, sqrt(0.5*(m_xx + m_yy))
 */
double Shape::getRms() const {
    double const T  = (_mxx + _myy);      // Trace
    return sqrt(0.5*T);
}

/**
 * Return e1's standard deviation
 */
double Shape::getE1Err() const {
    double const T  = (_mxx + _myy);      // Trace
    assert(T > 0);
    double const T4 = 4/(T*T*T*T);      // for our convenience
    
    double const e1e1Err = T4*(_covar(1, 1)*_myy*_myy + _covar(2, 2)*_mxx*_mxx - 2.0*_covar(1, 2)*_mxx*_myy);
    return sqrt(e1e1Err);
}

/**
 * Return sign(covar_{e1,e2})*sqrt(|covar_{e1,e2}|)
 */
double Shape::getE1E2Err() const {
    double const T  = (_mxx + _myy);      // Trace
    assert(T > 0);
    double const T4 = 4/(T*T*T*T);      // for our convenience
    
    double const var_e1e2 = T4*(-1.0*_myy*_mxy*_covar(1, 1) + _mxx*_mxy*_covar(2, 2) +
                                (_mxx - _myy)*_mxy*_covar(1, 2) + T*(_myy*_covar(1, 3) - _mxx*_covar(2, 3)));

    return (var_e1e2 > 0) ? sqrt(var_e1e2) : -sqrt(-var_e1e2);
}

/**
 * Return e2's standard deviation
 */
double Shape::getE2Err() const {
    double const T  = (_mxx + _myy);      // Trace
    assert(T > 0);
    double const T4 = 4/(T*T*T*T);      // for our convenience
    
    double const var_e2e2 = T4*(_mxy*_mxy*(_covar(1, 1) + _covar(2, 2) + 2.0*_covar(1, 2)) -
                                2.0*T*_mxy*(_covar(1, 3) + _covar(2, 3)) + T*T*_covar(3, 3));

    return sqrt(var_e2e2);
}

/**
 * Return the object's RMS size's standard deviation
 */
double Shape::getRmsErr() const {
    double const T  = (_mxx + _myy);      // Trace
    double const varT = _covar(1, 1) + _covar(2, 2) + 2*_covar(1, 2); // T's variance
    double const ms = 0.5*T;            // rms is sqrt(ms))
    double const varMs = 0.25*varT;   // ms's variance

    assert(ms > 0);
    return 0.5*sqrt(varMs/ms);          // error in sqrt(ms)
}
            
/************************************************************************************************************/
/**
 * @brief Call the concrete shape algorithm
 *
 * N.b. One purpose of this routine is to provide a place to specify default values for arguments
 */
template<typename ImageT>
Shape MeasureShape<ImageT>::apply(ImageT const& image, ///< The image containing the object
                                  double xcen,         ///< object's column position
                                  double ycen,         ///< object's row position
                                  PSF const* psf,      ///< image's PSF
                                  double background    ///< image's background level
                                  ) const {
    int const x = afwImage::positionToIndex(xcen);
    int const y = afwImage::positionToIndex(ycen);

    if (x - image.getX0() < 1 || x - image.getX0() > image.getWidth() - 2 ||
        y - image.getY0() < 1 || y - image.getY0() > image.getHeight() - 2) {
        throw LSST_EXCEPT(pexExceptions::RangeErrorException,
                          (boost::format("Object at (%.3f, %.3f) is too close to the edge of the frame") %
                           xcen % ycen).str());
    }
    pexLogging::TTrace<8>("meas.algorithms.shape", "Measuring shape of object at (%.3f, %.3f)", xcen, ycen);

    return doApply(image, xcen, ycen, psf, background);
}

/************************************************************************************************************/
/**
 * Return a MeasureShape of the requested variety
 *
 * @throws std::runtime_error if name can't be found
 */
template<typename ImageT>
MeasureShape<ImageT>* createMeasureShape(std::string const& name ///< desired variety
                                        ) {
    return MeasureShape<ImageT>::lookup(name).create();
}

/************************************************************************************************************/
//
// Explicit instantiations
// \cond
#define MAKE_SHAPEFINDERS(IMAGE_T) \
    template class MeasureShape<IMAGE_T>; \
    template MeasureShape<IMAGE_T>* \
    createMeasureProperty(std::string const&, IMAGE_T::ConstPtr, MeasureShape<IMAGE_T> const*)
                
MAKE_SHAPEFINDERS(lsst::afw::image::MaskedImage<float>);

// \endcond
                
}}}

File: tests/shape.py
#!/usr/bin/env python
# 
# LSST Data Management System
# Copyright 2008, 2009, 2010 LSST Corporation.
# 
# This product includes software developed by the
# LSST Project (http://www.lsst.org/).
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the LSST License Statement and 
# the GNU General Public License along with this program.  If not, 
# see <http://www.lsstcorp.org/LegalNotices/>.
#

import re, os, sys
import glob
import math
import unittest

import lsst.pex.policy as pexPolicy
import lsst.pex.exceptions as pexExceptions
import lsst.afw.image as afwImage
import lsst.afw.math as afwMath
import lsst.afw.geom as afwGeom
import lsst.afw.table as afwTable
import lsst.meas.algorithms as algorithms
import lsst.utils.tests as utilsTests
import lsst.afw.detection as afwDetection

import lsst.afw.display.ds9 as ds9

try:
    type(verbose)
except NameError:
    verbose = 0
    display = False

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

class ShapeTestCase(unittest.TestCase):
    """A test case for centroiding"""

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def testCleanup(self):
        """Test that tearDown does"""
        pass

    def do_testmeasureShape(self):
        """Test that we can instantiate and play with a measureShape"""

        algorithmName = "shape.sdss"
        algorithmConfig = algorithms.SdssShapeConfig()

        im = afwImage.ImageF(afwGeom.ExtentI(100))
        msk = afwImage.MaskU(im.getDimensions()); msk.set(0)
        var = afwImage.ImageF(im.getDimensions()); var.set(10)
        mi = afwImage.MaskedImageF(im, msk, var)
        del im; del msk; del var
        exp = afwImage.makeExposure(mi)
        
        #-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
        #
        # Add a Gaussian to the image
        #
        for a, b, phi, tol in [              # n.b. phi in degrees
            (2.5, 1.5, 90.0, 1e-3),
            (1.5, 2.5,  0.0, 1e-3),
            (1.5, 2.5, 45.0, 1e-3),
            (1.5, 2.5, 90.0, 1e-3),

            (3.0, 2.5,  0.0, 1e-3),

            (3.0, 12.5, 0.0, 1e-3),
            (3.0, 12.5, 0.0, 2e-4),

            (1.0,   1.0, 0.0, 4e-3),
            (1.0,  0.75, 0.0, 2e-2),
            #(0.75, 0.75, 0.0, 1e-1),
            ]:
            if b > a:
                a, b = b, a
                phi -= 90
            a, b, phi = float(a), float(b), math.radians(phi)

            im = mi.getImage()
            bkgd = 100.0
            x, y = 30, 40               # centre of object
            im[:] = bkgd
            
            axes = afwGeom.ellipses.Axes(a, b, phi*afwGeom.radians, True)
            quad = afwGeom.ellipses.Quadrupole(axes)
            if False:
                a0, b0 = a, b
                pixellatedAxes = axes.convolve(afwGeom.ellipses.Quadrupole(1/6.0))
                a, b = pixellatedAxes.getA(), pixellatedAxes.getB()
                print a, b, a0, b0
            sigma_xx, sigma_yy, sigma_xy = quad.getIxx(), quad.getIyy(), quad.getIxy()

            ksize = 2*int(4*max(a, b)) + 1
            c, s = math.cos(phi), math.sin(phi)

            sum, sumxx, sumxy, sumyy = 4*[0.0] if False else 4*[None]
            for dx in range(-ksize/2, ksize/2 + 1):
                for dy in range(-ksize/2, ksize/2 + 1):
                    u, v = c*dx + s*dy,  s*dx - c*dy
                    I = 1000*math.exp(-0.5*((u/a)**2 + (v/b)**2))
                    im[x + dx, y + dy] += I

                    if sum is not None:
                        sum += I
                        sumxx += I*dx*dx
                        sumxy += I*dx*dy
                        sumyy += I*dy*dy
                        
            if sum is not None:
                sumxx /= sum; sumxy /= sum; sumyy /= sum
                print "RHL %g %g %g" % (sumxx, sumyy, sumxy)

            algorithmConfig.background = bkgd
            schema = afwTable.SourceTable.makeMinimalSchema()
            shapeFinder = algorithms.MeasureSourcesBuilder()\
                .addAlgorithm(algorithmConfig.makeControl())\
                .build(schema)

            if display:
                ds9.mtv(im)

            table = afwTable.SourceTable.make(schema)
            table.defineShape(algorithmName)
            table.defineCentroid(algorithmName + ".centroid")
            source = table.makeRecord()
            center = afwGeom.Point2D(x, y)

            shapeFinder.apply(source, exp, center)

            if False:
                Ixx, Iyy, Ixy = source.getIxx(), source.getIyy(), source.getIxy()
                A2 = 0.5*(Ixx + Iyy) + math.sqrt( (0.5*(Ixx - Iyy))**2 + Ixy**2 )
                B2 = 0.5*(Ixx + Iyy) - math.sqrt( (0.5*(Ixx - Iyy))**2 + Ixy**2 )

                print "I_xx:  %.5f %.5f" % (Ixx, sigma_xx)
                print "I_xy:  %.5f %.5f" % (Ixy, sigma_xy)
                print "I_yy:  %.5f %.5f" % (Iyy, sigma_yy)
                print "A2, B2 = %.5f, %.5f" % (A2, B2)            

            self.assertTrue(abs(x - source.getX()) < 1e-4, "%g v. %g" % (x, source.getX()))
            self.assertTrue(abs(y - source.getY()) < 1e-4, "%g v. %g" % (y, source.getY()))
            self.assertTrue(abs(source.getIxx() - sigma_xx) < tol*(1 + sigma_xx),
                            "%g v. %g" % (sigma_xx, source.getIxx()))
            self.assertTrue(abs(source.getIxy() - sigma_xy) < tol*(1 + abs(sigma_xy)),
                            "%g v. %g" % (sigma_xy, source.getIxy()))
            self.assertTrue(abs(source.getIyy() - sigma_yy) < tol*(1 + sigma_yy),
                            "%g v. %g" % (sigma_yy, source.getIyy()))

    def testSDSSmeasureShape(self):
        """Test that we can instantiate and play with SDSSmeasureShape"""

        self.do_testmeasureShape()

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

def suite():
    """Returns a suite containing all the test cases in this module."""
    utilsTests.init()

    suites = []
    suites += unittest.makeSuite(ShapeTestCase)
    suites += unittest.makeSuite(utilsTests.MemoryTestCase)

    return unittest.TestSuite(suites)

def run(exit = False):
    """Run the tests"""
    utilsTests.run(suite(), exit)
 
if __name__ == "__main__":
    run(True)
